
## COMP5712M: Programming for Data Science


## Assignment 2: Cities and Earthquakes

### @author: Kevin Timothy Muller
#### Last Modified: 27th September 2023

### Do not import any other module besides those already provided in the notebook

## Question 1: World Cities

In this coursework exercise, you will download data from the provided link and read it in as a `CSV` file using the Pandas data analysis package for Python.
The data we will use contains a variety of information about cities from around the world.

### Coding Techniques for Working with `DataFrame`s

To complete these tasks you will need to access and filter a `DataFrame`. 
The `DataFrame` data structure has many convenient features for extracting and ordering information. Although conceptually it can be thought of as a comptuational
represention of a table, it is quite a complex data structure and takes a while
to master. The following questions can be done with only a small but powerful
set of `DataFrame` operations; and the following examples of typical forms of programming with `DataFrame`s should be useful 
for coding your answers.

#### Loading data from a CSV file into a `DataFrame`

`DataFrame`s are specifically designed to handle data organised in a tabular format.
Hence, as we would expect, since `CSV` is the standard format for tabular data, it is very easy to 
create a `DataFrame` by loading data from a `CSV` file.

* ```pandas.read_csv(source)``` ---
The  [read_csv function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html#pandas.read_csv) can accept a filename or URL as an argument.

#### Getting the Data for this Question

Download the data file ```worldcities.csv``` from Minerva, and put
the file in the same directory as this Jupyter notebook file.
Then, by running the following cell, we can set the 
global variable ```WC_DF``` to a DataFrame containing the information from ```worldcities.csv```. 

In [2]:
## WC_DF Initialisation

import pandas  ## This is the module for creating and manipulating DataFrames

WC_DF = pandas.read_csv("worldcities.csv")

**Be sure to keep the same variable name** `WC_DF` **for this global variable, otherwise most of the
following code will not work and you will break the autograder.**

#### Checking the contents of a DataFrame

Pandas provides the following useful methods that enable you to quickly check the contents of a `DataFrame`:

* ```df.head()``` ---
  For a DataFrame object, ```df```, this method extracts the first 5 rows of data, so you can easily check what the data looks like.
  
* ```df.describe()``` --- for a DataFrame, ```df```, this method provides a table giving and overview of some basic statistical properties of the DataFrame.

Note that the ```head()``` and ```describe()``` methods are actually operations that 
return a new DataFrame object. If this value is returned by the last line of a cell
it will be displayed as a table, but if it is generated elsewhere in the code
you will not see any output unless you use the ```display``` function from the
```IPython.display``` module.

#### Accessing DataFrame columns and rows

Each column of a `DataFrame` is a list-like object called a
`Series`. Elements, and slices of a `Series` can then be accessed in similar
fashion to a list. The following illustrates how get the `Series` containing
the first 5 elements of the `city` column of `WC_DF`:

In [3]:
top_5_cities = WC_DF["city"][:5]   ## selects the first 5 items of the "city" column.
top_5_cities

0          Tokyo
1       New York
2    Mexico City
3         Mumbai
4      São Paulo
Name: city, dtype: object

In the above output, the left hand column of the displayed value of `top_5_cities` 
shows the index label of each element. One of the differences between a `Series` and an ordinary list is that, whereas a list always has integers for its index labels, a `Series` can have different kinds of values for these. For instance (though there
is no reason to do this for the current assignment) we could set the index values to alphabetic letters, as follows:

In [4]:
top_5_cities.index = list("abcde")
top_5_cities

a          Tokyo
b       New York
c    Mexico City
d         Mumbai
e      São Paulo
Name: city, dtype: object

In [5]:
top_5_cities.index

Index(['a', 'b', 'c', 'd', 'e'], dtype='object')

You can also use ```.values``` to return an `array` of the column values without
the index:

In [6]:
WC_DF["city"][:5].values

array(['Tokyo', 'New York', 'Mexico City', 'Mumbai', 'São Paulo'],
      dtype=object)

An `array` is also a list-like datastructure. It does not have an index. The main difference between a list and an `array` is that the list is optimised for
storing large amounts of information and for efficiently applying numerical and
other operations to all elements of the array. Hence, `array`s are usually preferred
to lists when handling large amounts of information, or when storing numerical
vectors.

You can also easily find the column names of the DataFrame using ```.columns```, for example:

In [7]:
WC_DF.columns

Index(['city', 'city_ascii', 'lat', 'lng', 'country', 'iso2', 'iso3',
       'admin_name', 'capital', 'population', 'id'],
      dtype='object')

__Note:__ The `Index` returned here is yet another type of list-like, object. It is similar to an array,
except that it is used for indexing a `Series` or `DataFrame`. You do not usually
need to create or deal with `Index` objects directly, since this is done automatically when you create and minipulate `DataFrame`s. So you will normally only see one, when
you want to look at the columns or rows of a `DataFrame`. But what you should be
aware of, when dealing with `DataFrames`, is that the word _index_ can refer to
several different types of thing.

In many cases you can treat `Series`, `array` and `Index` objects like lists and if you want to change them to an ordinary list you can just use the `list` operator,
as in the following:

In [8]:
list(WC_DF.columns)

['city',
 'city_ascii',
 'lat',
 'lng',
 'country',
 'iso2',
 'iso3',
 'admin_name',
 'capital',
 'population',
 'id']

We can refer to rows of a `DataFrame` either by the expression `DF.loc[label]`, where `label` is the index label of the row we want, or by `DF.iloc[n]`,
where `n` is an `int` giving the position of the row in the `DataFrame`.
In the case of `WC_DF`, the labels are integers, so we would get the same result using either. You could test this. You could also see the difference if you try finding a row of `top_5_cities` `DataFrame` defined above, after its index labels have been replaced by letters. In this case you could access rows either using letters, using `loc`, or by `int`s, using `iloc`.

In [9]:
WC_DF.iloc[1]

city               New York
city_ascii         New York
lat                 40.6943
lng                -73.9249
country       United States
iso2                     US
iso3                    USA
admin_name         New York
capital                 NaN
population       19354922.0
id               1840034016
Name: 1, dtype: object

#### Iterrating through the rows of a DataFrame
A convenient way of going through the rows of a `DataFrame` to perform some operation i by using the `iterrows` method in a `for` loop. This enables you to get both the index label and the row itself, for each successive row of the `DataFrame`. The following code is a simple example:

In [10]:
for i, row in WC_DF.iterrows():
    print(i, row['city_ascii'], row['lat'], row['lng'])
    if i >3: break

0 Tokyo 35.685 139.7514
1 New York 40.6943 -73.9249
2 Mexico City 19.4424 -99.131
3 Mumbai 19.017 72.857
4 Sao Paulo -23.5587 -46.625


#### Sorting the rows of a DataFrame

It is easy, and often very useful, to sort the DataFrame by column values using ```.sort_values```, for example:

In [11]:
WC_DF.sort_values(by=["country"], ascending=True)[:10] # Sorts countries by alphabet

city       city_ascii      lat      lng      country iso2  \
10235             Karukh           Karukh  34.4868  62.5918  Afghanistan   AF   
8527   Kōṯah-ye ‘As̲h̲rō  Kotah-ye `Ashro  34.4500  68.8000  Afghanistan   AF   
3341          Shibirghān       Shibirghan  36.6580  65.7383  Afghanistan   AF   
6050               Khōst            Khost  33.3395  69.9204  Afghanistan   AF   
5141       Maḩmūd-e Rāqī    Mahmud-e Raqi  35.0167  69.3333  Afghanistan   AF   
2088         Lashkar Gāh      Lashkar Gah  31.5830  64.3600  Afghanistan   AF   
3210              Gardēz           Gardez  33.6001  69.2146  Afghanistan   AF   
6249        Maīdān Shahr     Maidan Shahr  34.3956  68.8662  Afghanistan   AF   
2105            Maīmanah         Maimanah  35.9302  64.7701  Afghanistan   AF   
7399             Andkhōy          Andkhoy  36.9317  65.1015  Afghanistan   AF   

      iso3 admin_name capital  population          id  
10235  AFG      Herāt   minor     17484.0  1004546127  
8527   AFG     Wardak     NaN     35008.0  1004450357  
3341   AFG    Jowzjān   admin     93241.0  1004805783  
6050   AFG      Khōst   admin         NaN  1004919977  
5141   AFG     Kāpīsā   admin      7407.0  1004151943  
2088   AFG    Helmand   admin    201546.0  1004765445  
3210   AFG    Paktiyā   admin    103601.0  1004468894  
6249   AFG     Wardak   admin         NaN  1004798735  
2105   AFG     Fāryāb   admin    199795.0  1004622920  
7399   AFG     Fāryāb   minor     71730.0  1004472345

##### Note on __encodings__ of the city name
there are two columns that hold the city name. The first column name is `'city'` and
the second is `city_ascii`. There are various different ways in which textual
information can be encoded into bytes. These days [Unicode characters](https://home.unicode.org/)
encoded using [UTF-8](https://en.wikipedia.org/wiki/UTF-8) are pretty standard.
But the older [ASCII](https://en.wikipedia.org/wiki/ASCII) code, which uses
a single byte per character is still commonly used. Unicode provides a huge
variaty of text characters and other symbols, whereas ASCII is quite 
limited (mainly to characters and symbols found in standard English). 
But ASCII and is simpler and in 
some ways easier to deal with than UTF-8. In the following questions you
will be asked to use the ASCII version of the city name (from the `city_ascii` column). This mainly just to make you aware that there are different encodings
of text strings, but it will also prevent cerain problems that could occur in the
Autograder, if different people used different encodings.

#### Filtering DataFrames
By _filtering_ we mean keeping some parts that we want and throwing away others.
Typically, we look for rows that match some condition; and the filter condition
is often some constraint involving the values for that row in one or more
columns.
`pandas` `DataFrame`s can  be filtered according values of a column by using a boolean expression, for example:

In [12]:
filtered_DF = WC_DF[ WC_DF['capital'] == 'admin'] # This keeps only administrative capitals
filtered_DF.head()

city city_ascii      lat       lng country iso2 iso3   admin_name  \
3     Mumbai     Mumbai  19.0170   72.8570   India   IN  IND  Mahārāshtra   
4  São Paulo  Sao Paulo -23.5587  -46.6250  Brazil   BR  BRA    São Paulo   
5      Delhi      Delhi  28.6700   77.2300   India   IN  IND        Delhi   
6   Shanghai   Shanghai  31.2165  121.4365   China   CN  CHN     Shanghai   
7    Kolkata    Kolkata  22.4950   88.3247   India   IN  IND  West Bengal   

  capital  population          id  
3   admin  18978000.0  1356226629  
4   admin  18845000.0  1076532519  
5   admin  15926000.0  1356872604  
6   admin  14987000.0  1156073548  
7   admin  14787000.0  1356060520

This way of filtering is a very powerful and useful aspect of `DataFrames`. 
However, the
syntax of the filter operation is rather unusual and a bit difficult to understand.

What is happening can be explained by these steps in the way a filter expression
is evaluated:
*  `DF['label']` (where `DF` is any `DataFrame`), gives a `Series` corresponding
    to    the `'label'` column of `DF.
    
*  `a_series == val` is a special use of `==`. When a Boolean operator
    (such `==`, `<` etc.) is applied to a `Series` object
   the result is actually a `Series` of Boolean values (not a single Boolean).
   The new `Series` obtained will have the value `True` for each element where
   the original `Series` satisfies `element == val`, and `False` for the rest.
   
* `DF[ bool_series ]`, is a special kind of slice-like operation, where a boolean
   series is given as a selection argument to the `DataFrame`. It will return 
   a new DF, containing all the rows of `DF` for which `bool_series` has the 
   value `True`. (These rows can be quickly found because the `DataFrame` and
   the Boolean series both have the same `Index`.)
   
You do not necessarily need to follow all of that precise desciption of filtering
but it will be extremely helpful if you are able to construct filtering
operations similar to the above example. You will see another example below,
in relation to the earthquake data you will be processing.

### Overview of Question 1 tasks

This question requires you to write functions to carry out the following specific tasks. 

* __Question 1a__ --- Find all city names that include *non-ASCII* characters. __[1 Mark]__

* __Question 1b__ --- Find city names that occur multiple times in the dataset. __[2 Marks]__
* __Question 1c__ --- Create a dictionary of the number of cities in each country that are included in the dataset. __[1 Mark]__

* __Question 1d__ --- Create a `DataFrame` of the largest cities (by population) in the world. __[2 Marks]__
* __Quesiton 1e__  --- Find all cities in a given country whose population is above a given number. __[2 Marks]__

* __Question 1f__ --- Find the total population of people living in cities in a given country. __[2 Marks]__

Full details for each task are explained below.

### Question 1a

As we have seen, the names of some cities include accents or symbols that are not represented in the standard ASCII character set. Write a function `non_ascii_cities`, which returns a `Set` containing all the cities that occur in the `world_cities.csv` dataset whose name **cannot** be properly represented using only ASCII characters. 

Your function should not have any arguments. You should assume that the global variable WC_DF has already been initialised by running the first code cell in this notebook (see above).

In [13]:
# Question 1a answer cell

def non_ascii_cities():
    non_ascii = []
    ASCII = '\x00\x01\x02\x03\x04\x05\x06\x07\x08\t\n\x0b\x0c\r\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\x7f'
    for i, row in WC_DF.iterrows():
        for ch in row['city'].upper():
            if ch not in ASCII:
                non_ascii.append(row['city'])
                break
    return set(non_ascii)

    # Modify to return a set of all non-ascii city names in the world_cities data

### Question 1b

One issue that you will discover if you investigate the `worldcities.csv` data is that there are many different cities that have the same name. You may find it interesting to discover which are the most common city names. But for this question you need to write a function `num_cities_occurring_n_times(n)`, such that:
    
* The argument `n` is an integer;
* The function returns an integer (type `int`), which is the number of different city names that occur `n` times in the `worldcities` dataset. 
* The function should consider the city name to be the value in the `city` column. In other words the name in the form that may contain non-ASCII characters.

In [14]:
# Question 1b answer cell

def num_cities_occurring_n_times(n):
    anslist = []
    for i, row in WC_DF.iterrows():
        if list(WC_DF['city']).count(row['city']) == n:
            anslist.append(row['city'])
    return len(set(anslist))
    # Modify to return a value according to the specification given above

### Question 1c

Write a function that returns a dictionary (a `dict` object), whose keys are all the country name strings that occur in the `worldcities` data and whose values are `int`s giving the number of cities of that country that are included in the dataset.



In [15]:
# Question 1c answer cell

def country_num_cities_dict():
    Dic = {}
    temp = {}
    for i, row in WC_DF.iterrows():
        if row["country"] not in list(Dic.keys()):
            temp = {row["country"] : (list(WC_DF['country']).count(row['country']))}
            Dic = {**Dic, **temp}
    return Dic

### Question 1d
Write a function `largest_cities_dataframe` that takes an `int` argument `n`  and uses the
pandas DataFrame `WC_DF` to return a new `DataFrame` containing `n` rows corresponding to
the `n` largest cities in terms of population size, in order of decreasing population size.

You should return a dataframe such that it has the same columns as the `WC_DF` and each
row has the same values as a corresponding row of `WC_DF`. It does not matter if the row
indexes are the same. (They may or may not be the same depending on the specific way
that you create the new DataFrame.)

In [17]:
# Question 1d answer cell

def largest_cities_dataframe(n):
    return WC_DF.sort_values(by=["population"], ascending=False)[:n]
    # Modify to return a dataframe of the n cities with the largest population

__NOTE:__ In answering __1d__ you may assume that no two cities have exactly the same population, which is almost but not quite certain, when dealing with large numbers like this. But, of course, when dealing with quantites where multiple data records could have the same value, we need to be careful, because this may not be the case.
For example, if we are interested in what equipement students own, we might think it would be informative
to find 'the top 10 students owning the most laptops'. In this case there could be: 1 student with 3 laptops, 23 students with 2 laptops, 160 with 1 laptop and 3 who do not own a laptop. In such a case it is not meaningful to pick the 'top 10' in terms of laptop ownership. A similar problem could potentiall occur with the earthquake data that we will look at later, because the earthquake magnitudes are only recorded to 1 decimal place.

### Question 1e
Define a function `big_cities_in_country( country, population)` that takes as arguments a string corresponding to the name of a country and an integer, which will referes to a population number.
The function should return a `list` of the form 
<center>
    <tt>[("city1", pop1), ("city2", pop2),... ]</tt>, 
</center>

where each pair
`("cityN", popN)` is a `tuple` consisting of the name, **in ASCII form**, of a city in the given `country`, followed by an `int`, which is the population of that city (according to the worldcities data). The list should include all and only those cities in the country whose population is greater than or equal to the given `popuplation` argument. The list should be ordered so that the `("cityN", popN)` items occur in **increasing** order of the population size `popN`.



In [33]:
# Complete question 1e in this cell

def big_cities_in_country(country, population): # country is a string argument
    df = WC_DF[WC_DF["country"]==country][WC_DF["population"]>=population][["city","population"]].sort_values(by=["population"], ascending=True)
    return list(df.itertuples(index=False, name=None))

### Question 1f
Create a function that given a country name, returns an `int` which is the total population of 
people liveing in all the cities of that country, as given in `WC_DF`. 

**Hints:** 
* You can ignore cities for which no population value is recorded in `WC_DF`. You will need to be able to deal with `NaN` values. A distinctive feature of `NaN` values is that they have been defined so that `x != x` has the value `True` if `x` has a `NaN` value. 

In [60]:
## Question 1f Answer Code Cell

def country_total_cities_population(country):
    return (int(sum((WC_DF[WC_DF["country"]==country][WC_DF["population"]==WC_DF["population"]]["population"]).values.tolist())))

### Question 2: Earthquakes - Web Access and Pandas DataFrames

In this coursework exercise, you will learn how to download live information
from the web and procress it using the Pandas data analysis package for Python.

The data we will use as an example is from the 
[United States Geological Survey (USGS)](https://www.usgs.gov/), which 
provides a wide range of geographic and geological information and data.
We shall be using their data relating to seismological 
events (i.e. Earthquakes) from around the world, which is published in the 
form of continually updated CSV files. Information about these feeds can
be found [here](https://earthquake.usgs.gov/earthquakes/feed/). The URL for the particular feed we shall be using is given below.

Questions Overview

* __Q2a__ --- Initialise a Pandas DataFrame by downloading earthquake data from the web. __[1 mark]__
* __Q2b__ --- Find earthquakes of a given magnitude or higher.           __[1 mark]__
* __Q2c__ --- Make a DataFrame showing the most powerful quakes          __[2 marks]__
* __Q2d__ --- Make a DataFrame showing distance of quakes from a given location __[3 marks]__
* __Q2e__ --- Identify all cities endangered by earthquakes   __[3 marks]__

### Question 2a: Read in data file
Read earthquake data from the USGS live feed CSV ```all_day.csv``` into a Pandas DataFrame.
The data can be obtained directly from  http://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/all_day.csv and read into a Pandas DataFrame.

__Note:__ For this question you do not need to download and save the file ```all_day.csv```. It should
be loaded directly from the web feed. However, while testing, if you have no internet connection or
a bad connection you could download a copy of the file. But remember to put it back to downloading
the current one before you submit. Note also that ```all_day.csv``` is a live file, which lists
quakes recorded during the past 24 hours, and is updated every minute, so of course,
you will not always get the same file or the same results. More information about this and other
earthquake feeds provided by USGS can be found [here](https://earthquake.usgs.gov/earthquakes/feed/v1.0/csv.php).

In [81]:
# Q2a answer code cell

import pandas    ## This is the module for creating and manupulating DataFrames

# Here we have assigned the url of the quake datasource to the global variable 
# 'QUAKE_SOURCE' for your convenience.
QUAKE_SOURCE = ( "http://earthquake.usgs.gov/" +
                 "earthquakes/feed/v1.0/summary/all_day.csv" )


QUAKE_DF = pandas.read_csv(QUAKE_SOURCE)  ## Modify this line to import the data using Pandas

#### You can use the following cell to test if you have read the quake data into `QUAKE_DF`

In [62]:
## If QUAKE_DF is a DataFrame, show the first 5 rows
try:
    if type(QUAKE_DF) == pandas.DataFrame:
        display(QUAKE_DF.head())
    else:
        print("QUAKE_DF is not a DataFrame")
except:
    print("QUAKE_DF has not been assigned a value")

time  latitude   longitude      depth   mag  \
0  2023-11-08T19:52:45.090Z   19.1705 -155.496170  34.119999  1.80   
1  2023-11-08T19:45:05.907Z   61.9622 -150.853900  67.500000  2.00   
2  2023-11-08T19:41:01.040Z   36.1605 -120.309166  10.200000  0.98   
3  2023-11-08T19:37:59.460Z   38.8480 -122.762337   0.940000  1.16   
4  2023-11-08T19:33:31.429Z   31.6480 -104.015000   4.705400  1.90   

      magType   nst    gap     dmin   rms  ...                   updated  \
0          md  27.0  209.0      NaN  0.09  ...  2023-11-08T19:55:55.830Z   
1          ml   NaN    NaN      NaN  0.31  ...  2023-11-08T19:47:21.629Z   
2          md   8.0  190.0  0.14490  0.02  ...  2023-11-08T19:42:38.585Z   
3          md   7.0  193.0  0.01223  0.03  ...  2023-11-08T19:49:10.918Z   
4  ml(texnet)  16.0   44.0  0.00000  0.10  ...  2023-11-08T19:51:41.149Z   

                         place        type horizontalError depthError  \
0   4 km SSW of Pāhala, Hawaii  earthquake            0.78   1.000000   
1  28 km E of Skwentna, Alaska  earthquake             NaN   0.700000   
2     5 km ENE of Coalinga, CA  earthquake            1.01   2.100000   
3          4 km NW of Cobb, CA  earthquake            0.61   0.740000   
4    39 km W of Mentone, Texas  earthquake            0.00   1.328851   

   magError  magNst     status  locationSource magSource  
0      0.59     3.0  automatic              hv        hv  
1       NaN     NaN  automatic              ak        ak  
2      0.04     2.0  automatic              nc        nc  
3      0.14     9.0  automatic              nc        nc  
4      0.20    13.0  automatic              tx        tx  

[5 rows x 22 columns]

##### Note:
The columns containing latitude and longitude values are labelled differently in the `worldcities.csv` and the earthquake data from USGS. This is a minor but very typical form of incompatibility between data formats that you will often need to deal with when working with real data.

### More examples of useful `pandas` functions

Here we show you some more pandas functions that you may find useful in this exercise. 

As we have seen, versatile filtering and sorting capabilities are provided by pandas. To get more understanding of these, you should look at tutorials of using Pandas DataFrames. But the following example illustrates how you can find and display quakes whose depth is greater than or equal to a given threshold:

In [63]:
def show_deep_quakes( depth ):
    # make deep_quakes DataFrame by selecting rows from QUAKE_DF
    deep_quakes = QUAKE_DF[ QUAKE_DF["depth"] >= depth ]  ## This is how you select rows by a condition
                                                          ## on one of the column values.
        
    print("Number of quakes of depth {} or deeper:".format(depth), 
           len(deep_quakes.index))     ## This finds the number of rows of the deep_quakes DataFrame
    
    display(deep_quakes.sort_values("depth", ascending=False))  ## Sort by descending depth value

**Note:**
The `QUAKES_DF` global variable needs to be set before these examples will work, so I am using a `try`, `except` construct to avoid getting an error.

In [64]:
try:
    show_deep_quakes(100)
except:
    print("Probably QUAKE_DF not correctly set")

Number of quakes of depth 100 or deeper: 7


time  latitude  longitude    depth  mag magType  \
197  2023-11-08T03:44:27.227Z  -23.3322  -179.5710  550.924  4.5      mb   
94   2023-11-08T12:28:16.882Z  -23.9740  -179.6124  477.355  4.5      mb   
65   2023-11-08T13:56:20.826Z   21.5708   143.0808  320.395  4.6      mb   
151  2023-11-08T10:02:02.233Z   63.0999  -150.8850  131.100  2.4      ml   
193  2023-11-08T04:13:01.334Z   60.0810  -153.0992  113.900  2.6      ml   
127  2023-11-08T10:59:43.504Z  -22.0908   -68.8081  111.064  4.4     mwr   
88   2023-11-08T12:39:41.924Z   37.0299    30.2316  100.993  4.4      mb   

       nst    gap   dmin   rms  ...                   updated  \
197   28.0  103.0  4.582  0.52  ...  2023-11-08T04:03:55.040Z   
94    43.0  100.0  4.939  1.10  ...  2023-11-08T12:42:10.040Z   
65   157.0   38.0  8.119  0.59  ...  2023-11-08T14:51:36.040Z   
151    NaN    NaN    NaN  0.30  ...  2023-11-08T10:17:31.040Z   
193    NaN    NaN    NaN  0.75  ...  2023-11-08T04:15:22.687Z   
127   17.0   46.0  0.878  0.70  ...  2023-11-08T11:10:03.040Z   
88    68.0   48.0  0.382  0.86  ...  2023-11-08T13:05:59.040Z   

                                        place        type horizontalError  \
197                 south of the Fiji Islands  earthquake           14.26   
94                  south of the Fiji Islands  earthquake           12.30   
65                     Mariana Islands region  earthquake            7.85   
151  64 km SE of Denali National Park, Alaska  earthquake             NaN   
193            65 km ENE of Pedro Bay, Alaska  earthquake             NaN   
127                42 km NNE of Calama, Chile  earthquake            5.62   
88                             Western Turkey  earthquake            6.09   

    depthError  magError  magNst     status  locationSource magSource  
197     12.428     0.113    23.0   reviewed              us        us  
94       9.338     0.086    39.0   reviewed              us        us  
65       5.099     0.035   248.0   reviewed              us        us  
151      0.700       NaN     NaN  automatic              ak        ak  
193      1.700       NaN     NaN  automatic              ak        ak  
127      8.695     0.080    15.0   reviewed              us        us  
88       6.924     0.103    27.0   reviewed              us        us  

[7 rows x 22 columns]

You can also find ```max``` and ```min``` values in a column. Eg:

In [65]:
try:
    QUAKE_DF["depth"].max()
except:
    print("Probably QUAKE_DF not correctly set")    

In [66]:
try:
    QUAKE_DF["depth"].min()
except:
    print("Probably QUAKE_DF not correctly set")    

### Question 2b: Find Powerful Quakes

Write a function `powerful_quakes` that takes a numerical argument and returns a `DataFrame` including
all the quakes in `QUAKE_DF` that have a magnitude greater than or equal to the given argument.

In [67]:
# Complete question 2b answer cell

def powerful_quakes(mag):
    return QUAKE_DF[QUAKE_DF["mag"]>=mag]

### Question 2c: Find `n+` most powerful earthquakes

Produce a `DataFrame` with rows represent the `n`(or maybe more) most powerful quakes
in descending order of magnitude. The returned ``DataFrame`` should show at least `n`
quakes and may sometimes show more since we do not want to leave out any quake that is equally
powerful as the last quake listed in the `DataFrame`,
More specificially, we want the function to return a `DataFrame` that:
* has exactly the same column names as `QUAKES_DF`,
* each row has the same vaues in each column as a corresponding row in `QUAKES_DF` (but it does not matter whether the row indices in the returned `DataFrame` are the same or different from corresponding rows in `QUAKES_DF`),
* the rows of the returned `DataFrame` are ordered in _descending order_ of their magnitude column value (rows of equal magnitude can appear in any order),
* contains all and only those rows `QUAKES_DF`, such that there are fewer than `n` other rows in
  `QUAKES_DF` that have a higher magnitude.


#### Note:
The above definition of the requirements is clear and precise. Though you may ask for help and advice regarding implementation, you will not be given help with understanding the specification. 

In [137]:
# Question 2c answer cell

def most_powerful_n_quakes(n):
    df = QUAKE_DF.sort_values(["mag"],ascending=False)
    c=0
    for i,row in df.iterrows():
        if(c==n-1):
            return df[df["mag"]>=int(row[["mag"]])]
        c = c+1
    # Edit this function to make it return a DataFrame of 
    # the 'top n' quakes of the all_day.csv file

### Distance between locations on the Earth's surface

Clearly, when dealing with data pertaining to locations in space, the distance between such locations is often of great significance when interpreting or extracting further information from the data.

To help answer the following questions you are provided with the function ```haversine_distance```, 
which implements the [_Haversine formula_](https://en.wikipedia.org/wiki/Haversine_formula) to find the surface distance in kilometres between two locations, that are specified in terms of
latitude and longitude values. When finding distances betwen points on the surface of the
Earth we need to use this formula, rather than the simpler Pythagorean distance formula,
because the Earth's surface is a sphere.

In [68]:
## Function to compute distance between locations (kilometres) 
# Returns the surface distance in meters, according to the Haversine formula,
# between two locations given as (latitude, longitude) coordinate pairs.

import math
def haversine_distance( loc1 , loc2 ): 
    '''finds the distance (m) between 2 locations, where locations are defined by
    longitudes and latitudes'''
    lat1, lon1 = loc1
    lat2, lon2 = loc2
    radius = 6371  # kilometers
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (math.sin(dlat / 2) * math.sin(dlat / 2) +
         math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) *
         math.sin(dlon / 2) * math.sin(dlon / 2))
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    d = radius * c
    return d

### Question 2d: Sort quakes by distance from a given location

Write a function `quake_distance_from_loc_dataframe(loc)` satisfying the following requirements:
* It takes a location argument `(latitude,longitude)` consisting of a pair of `float`s 
    (Note: this is a single argument but consists of a pair of values represented as Python `tuple`.)

* It returns a `DataFrame` object derived from `QUAKE_DF` but with one extra column `distance_from_loc`, giving the distance of each quake from the given location. 

* The rows of the returned `DataFrame` should be the same as those in `QUAKE_DF` except for the aditional `distance_from_loc` column. (However, it is not necessary to preserve the index of the `DataFrame`, this will be ignored when your solution is tested.)

* The rows of the returned `DataFrame` should be _sorted_ in order of _increasing_ values of `distance_from_loc`. 

* The original DataFrame `QUAKE_DF` should not be altered by the execution of 
`quake_distance_from_loc_dataframe(loc)`.

#### Note:
You will need to do some research to find out how to create a new column and set its values.

In [83]:
## 2d Answer Code Cell

def quake_distance_from_loc_dataframe(loc):
    column_values = []
    df = QUAKE_DF.copy()
    for i,row in QUAKE_DF.iterrows():
        column_values.append(haversine_distance((row["latitude"],row["longitude"]),loc))
    df["distance_from_loc"]=column_values
    df = df.sort_values(by=["distance_from_loc"], ascending=True)
    return df

### Question 2e: Identifying Endangered Cities

The idea of this question is to identify possible emergency situations by finding
cities that are likely to suffer from the effects of
an earthquake.

#### Effect of an qarthquake at a distance from its epicenter
The effect of an earthquake on a city or person will depend on their distance from the source of the quake. The effect of an earthquake will depend on many factors and even the dependence on distance to source is very complex. However, after a bit of background research, Brandon has come up with a simple formula which hopefully at least gives a very crude estimate of relative effect of a quake with a particular magnitude and depth on a surface location at a known surface distance from the quake's epicenter. The calculated effective magnitude of an earthquake will be less that the source magnitude, for instance a magnitude 9 quake at a depth of 100km (which is likely to be extremely destructive), would have an effective magnitude of 5 at its epicentre (directly above the source) and 3.585 at a point on the earth surface 500km away from the epicenter.

In [90]:
def effective_magnitude( magnitude, depth, surface_distance ):
    energy = 10**magnitude  # convert logarithmic magnitude to a linear energy value
    if depth < 1:   # Crude fix for small or negative depths (can occur where land is above sea level)
        depth = 1
    ## Calculate distance to source by Pythagorus (ignoring curvature of surface)
    dist_to_source_squared =  depth**2 + surface_distance**2
    ## Apply inverse square distance multiplier to get energy density at distance from source
    ## (Ignores damping effects)
    attenuated_energy = energy/dist_to_source_squared
    attenuated_magnitude =  math.log10(attenuated_energy) ## Convert back to a log base 10 scale
    return attenuated_magnitude

# Some test cases.
#effective_magnitude(9,100,500)
#effective_magnitude(6,50, 100)

The _epicenter_ of an earthquake is the point on the earth that is directly above its source. Thus the effective magnitude at the epicenter is just the effective magnitude of the quake at surface distance zero:

In [91]:
def epicenter_magnitude( magnitude, depth ):
    return effective_magnitude( magnitude, depth, 0)

**Note:** For any given quake, its `effective_magnitude` at any point on Earth is always less than or equal to its `epicenter_magnitude`.

### Specification of the `endangered_cities` function

Now we get to the specification of the function.

Write a function `endangered_cities( min_population, min_effective_magnitude)`
that takes 
two numerical arguments: an `int` (`min_population`) and a `float` (`min_effective_magnitude`) 
and  returns a `list` specifying all those cities listed in 
the `WC_DF` such that:
* the city has  a population greater than or equal to the given `min_population`;

* for at least one of the quakes recorded in `QUAKE_DF`, the `effective_magnitude` (as determined by the
  function defined above) of that quake
   at the location of the city (as given in `worldcities.csv`) is equal to or greater than the  `min_effective_magnitude`.
   
* each city in the list should be represented by a tuple `(city_ascii, country, (lat, lng))`, giving
the city name in ASCII form, the country and the location (as a latitude, longitude pair).

* the returned list should be ordered aphabetically, primarily in terms of `country` and secondarily
  cities of the same country should be ordered in terms of `city_ascii`. 
  This ordering is illustrated by the sample output given below.

* A final condition is that the function should run in a reasonable time frame such that it can be expected to give correct and complete results after no more that 5 minutes excecution time. To ensure that this is feasible, the example cases you will be tested on, will be ones for which Brandon's solution ran in less than 2 minues on the Gradescope autograder platform. (All submissions will be tested on the same saved copy of `all_day.csv` previously downloaded from the USGS feed.)
   
##### Example Output:   

<pre>
 in [165]    %time        
             endangered_cities(200000, 0.5)
 
 out[165]    CPU times: user 1min 58s, sys: 768 µs, total: 1min 58s
             Wall time: 1min 58s
             [('Baghlan', 'Afghanistan', (36.1393, 68.6993)),
              ('Kunduz', 'Afghanistan', (36.728, 68.8725)),
              ('Mazar-e Sharif', 'Afghanistan', (36.7, 67.1)),
              ('Ambon', 'Indonesia', (-3.7167, 128.2)),
              ('Denov', 'Uzbekistan', (38.2772, 67.8872))]
</pre>

##### Notes:

* Your code should make use of the functions defined above: `haversine_distance`, `effective_magnitude` and `epicenter_magnitude`. You are advised not to change these otherwise you may get different results from what the Autograder is expecting.

* The calculation of this function is computationally intensive. For prelimiary testing you might use smaller data sets by, say only using cities in one country. 

* There are many optimisations that can be done to reduce the computational cost of finding the endangered cities. For example, calculating `epicenter_magnitude` enables weaker earthquakes to be ignored without the need to determine their distance from every city on earth. It may also be useful to not that when calcuating the effective magnitude of a given quake (with a specific depth) at different locations, the value always decreases as surface distance from the quake increases. 

* Remember that the autograder will run your actual code, so will probably take a while to grade this function. However, it only runs the function once and then performs 3 different tests on the value that is returned.

* The autograder will run on a presaved set of quakes (so the test is the same for all submissions). However, your code should run on any version of `all_day.csv` that you download from the USGS live feed.

In [118]:
## 2e Answer Code Cell
def endangered_cities(min_population, min_effective_magnitude):
    df1 = WC_DF[WC_DF["population"]>=min_population]
    df2 = pandas.DataFrame(columns = ["latitude","longitude","depth","mag"])
    preans = pandas.DataFrame(columns = ["city_ascii","country","lat","lng","id"])
    ans = []
    c=0
    for i,row in QUAKE_DF.iterrows():
        if epicenter_magnitude( row["mag"], row["depth"] ) >= min_effective_magnitude:
            df2.loc[len(df2)] = row
            c = c+1
    c=0
    for i, row in df2.iterrows():
        for j, row2 in df1.iterrows():
            if row2["id"] not in preans["id"]:
                if effective_magnitude(row["mag"], row["depth"], haversine_distance((row["latitude"],row["longitude"]), (row2["lat"],row2["lng"])))>= min_effective_magnitude:
                    preans.loc[len(preans)] = row2[["city_ascii","country","lat","lng","id"]]
                    c = c+1
    preans["comb"] = preans["country"] +  preans["city_ascii"]
    preans = preans.sort_values(by=["comb"], ascending=True)
    preans = preans.drop(columns = ["id","comb"])
    for i, row in preans.iterrows():
        ans.append((row["city_ascii"],row["country"],(row["lat"],row["lng"])))
    return ans